[clang-format] Allow custom pointer/ref alignment in return types

daandemeyer opened the pull request:
I would like to adopt clang-format in systemd
(https://github.com/systemd/systemd). One major blocker is that
systemd right-aligns pointers by default, except in function return
types, where they are left-aligned.

Let's introduce a ReturnType customization for PointerAlignment and
ReferenceAlignment that allows overriding the alignment used in return
types.

Fixes https://github.com/llvm/llvm-project/issues/136597

daandemeyer committed:
[clang-format][NFC] Upgrade PointerAlignment option to a struct

This allows adding other suboptions e.g. ReturnType for #136597.

daandemeyer committed:
[clang-format] Allow custom pointer/ref alignment in return types

I would like to adopt clang-format in systemd
(https://github.com/systemd/systemd). One major blocker is that
systemd right-aligns pointers by default, except in function return
types, where they are left-aligned.

Let's introduce a ReturnType customization for PointerAlignment and
ReferenceAlignment that allows overriding the alignment used in return
types.

systemd style guide: https://github.com/systemd/systemd/blob/main/docs/CODING_STYLE.md

Fixes #136597

HazardyKnusperkeks commented on clang/include/clang/Format/Format.h:
```diff
@@ -4017,6 +4017,27 @@ struct FormatStyle {
+    SAS_Right,
+    /// Align pointer/reference in the middle.
+    /// \code
+    ///   int *a(void);
```
You mixed up the examples.

HazardyKnusperkeks commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -6540,12 +6540,44 @@ void TokenAnnotator::printDebugInfo(const AnnotatedLine &Line) const {
 }
 
+static bool isReturnType(const FormatToken &Tok) {
+  // Look forward to see if there's a function declaration paren
```
```suggestion
  // Look forward to see if there's a function declaration paren.
```
Comments end in full stop.

HazardyKnusperkeks commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -6540,12 +6540,44 @@ void TokenAnnotator::printDebugInfo(const AnnotatedLine &Line) const {
+static bool isReturnType(const FormatToken &Tok) {
+  // Look forward to see if there's a function declaration paren
+  const FormatToken *Next = Tok.Next;
+  while (Next) {
```
Use a `for` loop.

HazardyKnusperkeks commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -2296,6 +2299,125 @@ TEST_F(FormatTest, SeparatePointerReferenceAlignment) {
+  verifyFormat("int** const f10(int **const a);", Style);
+  verifyFormat("const int** const f11(const int **const a);", Style);
+
+  // Triple pointers with const
```
Do you really think you need that many tests?

DaanDeMeyer replied:
Tests are cheap in this case so I figure why not have a ton

HazardyKnusperkeks reviewed (commented):
Can you link to a style guide?

daandemeyer committed:
[clang-format] Allow custom pointer/ref alignment for C-style casts

I would like to adopt clang-format in systemd
(https://github.com/systemd/systemd). One major blocker is that
systemd right-aligns pointers by default, except in function return
types and c-style casts, where they are left-aligned.

Let's introduce a CStyleCast customization for PointerAlignment and
ReferenceAlignment that allows overriding the alignment used for
C-style casts.

systemd style guide: https://github.com/systemd/systemd/blob/main/docs/CODING_STYLE.md

DaanDeMeyer commented:
> > Can you link to a style guide?
> 
> Can you link to a style guide?
> 
> For the review process is it better to just add commits and not amend them. On merge they will all be squashed into one commit.

systemd style guide: https://github.com/systemd/systemd/blob/main/docs/CODING_STYLE.md

HazardyKnusperkeks commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -2296,6 +2296,304 @@ TEST_F(FormatTest, SeparatePointerReferenceAlignment) {
 }
 
+TEST_F(FormatTest, ReturnTypeAlignment) {
+  // Test that ReturnType overrides PointerAlignment for return types
```
```suggestion
  // Test that ReturnType overrides PointerAlignment for return types.
```

HazardyKnusperkeks reviewed (commented):
Apart from that look okay to me.

DaanDeMeyer commented on clang/docs/ClangFormatStyleOptions.rst:
```diff
@@ -5668,31 +5668,107 @@ the configuration (without a prefix: ``Auto``).
+  Acceptable values (configured as a single string or with suboptions):
+  * ``Left``
+  * ``Right``
+  * ``Middle``
```
@mydeveloperday I'm not sure I follow, do you suggest `PointerAlignment: Custom` and then another option `PointerAlignmentOptions` struct to configure the individual options?

I'm fine to rework it like that if that's what you want for consistency but I also have to mention that I find my approach much nicer.

DaanDeMeyer commented:
> I'm open to adding options that make sense, but I don't see the rationale behind having different pointer alignments in function declarations ( e.g. `int* foo(int *)`) and type casts (e.g. `(int*)`).

For variable declarations, the pointer is semantically attached to the variable name as the variable name is a pointer to a value of the given type, so we align the pointer right for variables in systemd. It also makes much more sense to align right when you declare multiple pointer variables on the same line, which is something we do in systemd.

For function return types and c style casts, there is no variable name to attach the pointer to. And attaching the pointer to the function name doesn't make much sense, as the functionis not a pointer to a value, but returns a pointer to a value instead. So we decided that for these cases, it makes more sense to align the pointer left.

In the end all of this is bikeshedding though. Practically we can't adopt clang-format in systemd unless it formats code mostly like our style guide, and this feature is required to allow clang-format to adhere to our style guide.

DaanDeMeyer commented:
Ping @owenca

owenca commented:
I'm not convinced that we should move forward with this. Let's wait for @mydeveloperday.

mydeveloperday reviewed (changes requested):
I don't feel like my previous comments were addressed

DaanDeMeyer commented:
> I don't feel like my previous comments were addressed 

Because you're not replying to my questions. I clearly stated why I don't think what you suggested (or at least how I interpreted it) won't work and why we need Default.

Please take a look at my replies to your comments and answer my questions and I will take a look at addressing your comments.

mydeveloperday reviewed (commented):
I just think if we are introducing a struct mechanism for the option it should follow the previous struct pattern not make up a new one (Irrespective of if you feel its better, I think that subjective)... I'm still struggling with why we need this and not do it like we have with the BraceOption etc in the past. I'm a no to this design, but if @HazardyKnusperkeks  and @owenca  both agree its ok then I'll yield, but I'm not comfortable with it like this but I won't block them.
